### PR TITLE
Zephyr coap client

### PR DESCRIPTION
**Pull request by jarlamsa**
Add an asynchronous coap client.
The coap client processes requests, sends those to server and awaits for response.
Currently supports 1 request at a time.

### CODE REVIEW

**Comment by jarlamsa** [comment]
@SeppoTakalo @juhaylinen @plskeggs please review.

File: include/zephyr/net/coap_client.h
Comment (SeppoTakalo) [inline include/zephyr/net/coap_client.h:14]:
```suggestion
 * @param offset Payload offset.
```

File: include/zephyr/net/coap_client.h
Comment (SeppoTakalo) [inline include/zephyr/net/coap_client.h:10]:
```suggestion
 * @brief Callback for CoAP request.
 *
 * This callback is called for responses to CoAP client requests.
 * It is used to indicate errors, response codes from server or to deliver payload.
 * Blockwise transfers cause this callback to be called sequentially with increasing payload offset and only  
 * partial content in buffer pointed by @ref payload parameter.
```

File: include/zephyr/net/coap_client.h
Comment (SeppoTakalo) [inline include/zephyr/net/coap_client.h:13]:
```suggestion
 *                    @ref enum coap_response_code for positive.
```

Be systematic with using either `\` or `@` to indicate Doxygen command.

File: include/zephyr/net/coap_client.h
Comment (SeppoTakalo) [inline include/zephyr/net/coap_client.h:15]: What if there is no payload?
Is this `NULL` or how do we know the pointer is valid.

File: include/zephyr/net/coap_client.h
Comment (SeppoTakalo) [inline include/zephyr/net/coap_client.h:36]:
```suggestion
struct coap_client_request {
	enum coap_method method;      /**< Method of the request */
	bool confirmable;	      /**< CoAP Confirmable/Non-confirmable message */
	const char *path;	      /**< Path of the requested resource */
	enum coap_content_format fmt; /**< Content format to be used */
	uint8_t *payload;	      /**< User allocated buffer for send request */
	size_t len;		      /**< Length of the payload */
	coap_client_response_cb_t cb; /**< Callback when response received */
	void *user_data;	      /**< User provided context */
};
```

Please align the comment blocks. Also you need to use `/**<` when the Doxygen block refers to the previous member.

File: subsys/net/lib/coap/CMakeLists.txt
Comment (pdgendt) [inline subsys/net/lib/coap/CMakeLists.txt:8]: This is conditional based on `CONFIG_COAP_ASYNC_CLIENT`.

File: subsys/net/lib/coap/coap_client.c
Comment (pdgendt) [inline subsys/net/lib/coap/coap_client.c:38]: Making everything `static` like this only allows for a single client, can't we introduce a `struct coap_client` type that would allow for re-usability?

File: tests/net/lib/coap_client/CMakeLists.txt
Comment (pdgendt) [inline tests/net/lib/coap_client/CMakeLists.txt:25]: Why aren't these set in `prj.conf`?

**Review by pdgendt — DISMISSED** [review]
I think this needs more thought before being merged.

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [inline subsys/net/lib/coap/coap_client.c:625]: "There are more messages"

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [inline subsys/net/lib/coap/coap_client.c:497]: maybe `goto fail` instead of returning here?

**Review by plskeggs — COMMENTED** [review]
This is a good start for use cases where a single request/response at a time is all that is needed.  

I wonder if there are use cases for low power IoT devices which would want to POST or PUT multiple items (for example, sensor data readings), and not wait for ACK between each one, but instead handle that at the end.  Overlapping transfers could save battery life in that situation.

File: subsys/net/lib/coap/coap_client.c
Comment (rlubos) [reply subsys/net/lib/coap/coap_client.c:38]: I totally agree here, we've already been through this with other helper libs (like TFTP) - we'll eventually end up reworking the API so support multiple parallel clients.

IMO it'd be better to do it right from start, i. e. introduce a client context structure and provide it as an argument to the API functions.

File: subsys/net/lib/coap/coap_client.c
Comment (rlubos) [inline subsys/net/lib/coap/coap_client.c:406]: nit: No need for an empty line in such cases.

File: subsys/net/lib/coap/coap_client.c
Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c:38]: Good point, needs a bit of refactoring. In that case I think we could also drop the KConfig flag for `CONFIG_COAP_ASYNC_CLIENT`

File: tests/net/lib/coap_client/CMakeLists.txt
Comment (jarlamsa) [reply tests/net/lib/coap_client/CMakeLists.txt:25]: To isolate the tests from other parts of code and omit some of the KConfig dependencies.

File: subsys/net/lib/coap/coap_client.c
Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c:495]: That's correct, need to set the code as 0 in case of a separate confirmable response.

File: subsys/net/lib/coap/coap_client.c
Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c]: I made the changes now, the client uses `struct coap_client` which needs to be initialized with `coap_client_init` before using. This allows using multiple clients and nothing should be statically allocated in the client code now.

File: subsys/net/lib/coap/coap_client.c
Comment (pdgendt) [inline subsys/net/lib/coap/coap_client.c:652]: Turn the priority into a `kconfig` option.

File: include/zephyr/net/coap_client.h
Comment (pdgendt) [inline include/zephyr/net/coap_client.h:38]: This structure is used in the public part of the API, it should not be hidden.

File: include/zephyr/net/coap_client.h
Comment (pdgendt) [inline include/zephyr/net/coap_client.h:53]: The current implementation has a stack per CoAP client, if I were to do multiple requests at the same time, I either need to do these sequentially or create a client for each request, which is very costly.
This becomes even more apparent if I need to observe multiple server resources (long lasting requests).

It would be great if a single client could support multiple requests.
I think this is doable if you would move some members from `struct coap_client` to `struct coap_client_req` to keep track if a single request state. And let the client keep track of al ongoing requests (for example in a linked list).

Comment (jarlamsa) [reply include/zephyr/net/coap_client.h:53]: The multiple request is planned also, and doesn't need changes to the public API.

File: include/zephyr/net/coap_client.h
Comment (jarlamsa) [reply include/zephyr/net/coap_client.h:38]: True, but user should still consider it as opaque. Is there some guideline for the doxygen regarding this kind of case?

File: subsys/net/lib/coap/coap_client.c
Comment (rlubos) [inline subsys/net/lib/coap/coap_client.c:325]: Why the wait?

File: subsys/net/lib/coap/coap_client.c
Comment (rlubos) [inline subsys/net/lib/coap/coap_client.c:489]: `Set a timeout value` where exactly?

File: subsys/net/lib/coap/coap_client.c
Comment (rlubos) [inline subsys/net/lib/coap/coap_client.c:508]: Not sure I understand the logic behind this if here.

File: include/zephyr/net/coap_client.h
Comment (jarlamsa) [reply include/zephyr/net/coap_client.h:22]: Good point, it is mentioned in the description, but needs clarification: Blockwise transfers cause this callback to be called sequentially with increasing payload offset.

File: subsys/net/lib/coap/coap_client.c
Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c:508]: `client->pending.timeout` can be used in this case to identify if the response is a separate response, as the pending has been cleared and we are using a separate timeout value for the separate response.

File: include/zephyr/net/coap_client.h
Comment (SeppoTakalo) [inline include/zephyr/net/coap_client.h:88]:
```suggestion
 * or a background thread. Socket must be open and connected to a destination
 * address. Once the callback is called with result code 0, socket can be closed or
 * used for another query.
```

Is this contradicting the API a bit?
If socket is connected, why do we need the `sockaddr` for?

And when DTLS socket, which is already connected and do not accept any other socket addresses, should we allow `addr` to be NULL?

Comment (jarlamsa) [reply include/zephyr/net/coap_client.h]: True, latest changes allow now the `addr` to be NULL. Changed the documentation to reflect that.

**Review by SeppoTakalo — DISMISSED** [review]

File: subsys/net/lib/coap/coap_client.c
Comment (rlubos) [reply subsys/net/lib/coap/coap_client.c:560]: But isn't `COAP_TYPE_CON` response **always** a separate response? Otherwise it'd be piggybacked ACK. And the ACK for the separate response should be empty, `COAP_RESPONSE_CODE_OK` is a response code (and we're not replying for a request here, just acknowledging).

File: include/zephyr/net/coap_client.h
Comment (rlubos) [reply include/zephyr/net/coap_client.h:63]: Any feedback on this?

Comment (jarlamsa) [reply include/zephyr/net/coap_client.h:63]: Yes, will be done in the future, as well as the multi-request version of the client.

File: subsys/net/lib/coap/coap_client.c
Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c:560]: Actually the server may respond to a `COAP_TYPE_NON` with `COAP_TYPE_CON`, and as the server must not respond with `COAP_TYPE_ACK` in that case, the `COAP_TYPE_CON` answer isn't a separate message.

Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c:560]: https://www.rfc-editor.org/rfc/rfc7252#page-12

Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c:560]: Yeah, that's actually true, https://www.rfc-editor.org/rfc/rfc7252#section-5.2.2 states that also CON responses to NON -requests are also thought as separate responses. I'll make the change here.

Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c]: Done

**Review by rlubos — DISMISSED** [review]

File: include/zephyr/net/coap_client.h
Comment (pdgendt) [reply include/zephyr/net/coap_client.h:63]: I would've liked to have this functionality in this PR, as users will need to update/refactor their code in the future.

Comment (rlubos) [reply include/zephyr/net/coap_client.h:63]: Wouldn't such a change be transparent for the applications? Yes, the application allocates the client context structure, but it's internals should not be used by the app (`coap_client_init()` configures the thread).

File: include/zephyr/net/coap_client.h
Comment (plskeggs) [inline include/zephyr/net/coap_client.h:6]: Missing header guard

**Review by plskeggs — CHANGES_REQUESTED** [review]
Add header guard

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [inline subsys/net/lib/coap/coap_client.c:331]: This should be `client->socklen` not `sizeof(client->address)` here and everywhere below.

Otherwise, one cannot call `coap_client_req` with the `addr` parameter set to NULL as you intend based on line 299 above.

Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c]: Good catch, fixed now.

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [inline subsys/net/lib/coap/coap_client.c:41]: Note that `sendto` returns a positive value indicating the number of bytes sent.  This means that whatever calls `send_request` needs to know that.

File: include/zephyr/net/coap_client.h
Comment (plskeggs) [inline include/zephyr/net/coap_client.h:99]: As implemented, it returns a positive value indicating the number of bytes sent, not 0, on success.

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [inline subsys/net/lib/coap/coap_client.c:543]: It would be useful for the user of this library to get a callback when an ACK is received in response to a CON request that carries no additional payload (such as a PUT or POST).  Otherwise, the user does not know if the server accepted the request.

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [inline subsys/net/lib/coap/coap_client.c:600]: `result_code` is a duplicate of `response_code` -- they are both set to `coap_header_get_code(response)`.

File: subsys/net/lib/coap/coap_client.c
Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c:543]: But it might be confusing for the user if the server responds with ACK, but doesn't send response ever if the separate response takes too long.

File: include/zephyr/net/coap_client.h
Comment (jarlamsa) [reply include/zephyr/net/coap_client.h:99]: Changed to not return the number of bytes sent.

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [reply subsys/net/lib/coap/coap_client.c:545]: OK, sounds good.  It is important for the user to get a callback if the separate response never arrives, and retries are exhausted.

**Comment by plskeggs** [comment]
@jarlamsa -- I've been testing this out and noticed that you do not send a Content-Type header, only Accept.

It would be very helpful for compatibility with the server I am working with to be able to include both, ideally with separately specified values in the `coap_client_request` structure.

Optionally, being able to only set one, the other, or both, depending on the method (GET, PUT, etc.) and payload presence would be a bonus.

**Comment by jarlamsa** [comment]
> @jarlamsa -- I've been testing this out and noticed that you do not send a Content-Type value (`COAP_OPTION_CONTENT_FORMAT`), only Accept (`COAP_OPTION_ACCEPT`).
> 
> It would be very helpful for compatibility with the server I am working with to be able to include both, ideally with separately specified values in the `coap_client_request` structure. Optionally, being able to only set one, the other, or both, depending on the method (GET, PUT, etc.) and payload presence would be a bonus.
> 
> Unfortunately, the way the code is structured, `coap_client_init_path_options()` (not `coap_client_init_request()`) would need to handle the `COAP_OPTION_CONTENT_FORMAT` due to the requirement to add options in numeric order. So that makes this a bigger change than it could be. Maybe that function could simply return pointers to the start and end of the URI vs. query portions, perhaps iteratively, and let the calling function add the options to the packet, rather than do all of this in one function.

Thanks for noticing, yeah, the option ordering is a bit awkward in this case. However, CoAP-library now supports the option reordering, so as if you add an option too late, it will move it and reorganize the message.

**Review by SeppoTakalo — DISMISSED** [review]

**Comment by martinjaeger** [comment]
Nice work!

Could you please add a reference to this new API to the docs, so that the Doxygen comments are rendered and people are aware that the CoAP client exists?

**Comment by jarlamsa** [comment]
> Nice work!
> 
> Could you please add a reference to this new API to the docs, so that the Doxygen comments are rendered and people are aware that the CoAP client exists?

Yes, I plan to do that in a separate PR and add also similar usage example as is done with the HTTP-client.

**Review by plskeggs — CHANGES_REQUESTED** [review]

File: include/zephyr/net/coap_client.h
Comment (SeppoTakalo) [reply include/zephyr/net/coap_client.h:42]: This enum does not have a field for "ignore" or "unknown" so having it here forces Accept option to be always in the query.
In my mind, this feels a bit overkill. I have not yet seen a CoAP service that uses Accept option to change the content format. Most have hard coded.

Is this really necessary?

Comment (plskeggs) [reply include/zephyr/net/coap_client.h:42]: It may be overkill in some sense, but the standard does allow for it.  I know there is no point in implementing every possible option though.

That's a good point that there is no ignore or unknown value.  It would be nice to be able to omit either Accept or Content-Type.  Perhaps add such a value to the enum?  Or add another value to control which options to include, for example, a bitmask?

Using Accept on responses is currently necessary for the server we are developing, but I will discuss offline and respond later.

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [inline subsys/net/lib/coap/coap_client.c:588]: Is there a reason the subsequent block sizes are fixed at `COAP_BLOCK_128`?  What about either matching the size used in the first block, or the equivalent of MAX_COAP_MSG_LEN?

Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c:588]: Actually the `coap_update_from_block` after this updates the block context according to the block.

File: include/zephyr/net/coap_client.h
Comment (jarlamsa) [reply include/zephyr/net/coap_client.h]: Added 2 new fields. A list of extra options and a number of options accompanying it. This should now satisfy also the possible future requirements for extra options. Also added a KConfig for defining header size. If adding lots of extra headers, users should increase the header size as well to make sure the header and payload will fit to the buffer.

**Comment by jarlamsa** [comment]
> Nice work!
> 
> Could you please add a reference to this new API to the docs, so that the Doxygen comments are rendered and people are aware that the CoAP client exists?

I was planning on adding the documentation to different PR, but as there was still some changes, I pushed the documentation to this PR as well.

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]

**Comment by jarlamsa** [comment]
Thank you @SeppoTakalo for your review comments, I have addressed those now. Please re-review.

**Review by SeppoTakalo — APPROVED** [review]

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [inline subsys/net/lib/coap/coap_client.c:124]: `i` is undefined

**Review by plskeggs — APPROVED** [review]

File: subsys/net/lib/coap/coap_client.c
Comment (jarlamsa) [reply subsys/net/lib/coap/coap_client.c:124]: it is defined in the beginning of the function.

**Comment by jarlamsa** [comment]
@rlubos please re-review.

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [reply subsys/net/lib/coap/coap_client.c:124]: sorry, I placed this at the wrong point in the code

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [inline subsys/net/lib/coap/coap_client.c:234]: This is where I meant to comment about `i`, but I can see it defined above now on line 196.

When I made the comment earlier, I received a compile error on your code until I defined `i`.  Strange.

**Review by rlubos — APPROVED** [review]

**Comment by jarlamsa** [comment]
This now has 2 approvals with write access, could we get this merged in? @jukkar

**Pull request merged**
